File: routers/record.go
package routers

import (
	"book-keeping-web/models"
	"book-keeping-web/utils"
	"time"

	"github.com/gin-gonic/gin"
)

type RecordReq struct {
	ClassificationId    int    `json:"classification_id" binding:"required"`      // 分类id
	SubClassificationId int    `json:"sub_classification_id" binding:"omitempty"` // 子类别id
	Note                string `json:"note" binding:"omitempty"`                  // 备注
	Type                string `json:"type"`                                      // 类型
	Amount              string `json:"amount" binding:"required"`                 // 金额
	ConsumptionTime     string `json:"consumption_time" binding:"required"`       // 消费时间
}

// RecordCreate 为当前用户创建分类
func RecordCreate(c *gin.Context) {
	req := RecordReq{}
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.FailMessage(c, "parse param error")
		return
	}
	// 解析时间
	consumptionTime, err := time.Parse(utils.TimeLayout, req.ConsumptionTime)
	if err != nil {
		utils.FailMessage(c, "parse consumption_time error.")
		return
	}
	record := models.Record{Amount: req.Amount, UserId: c.GetInt("UserId"), ClassificationId: req.ClassificationId,
		Note: req.Note, Type: req.Type, ConsumptionTime: consumptionTime, SubClassificationId: req.SubClassificationId}
	if !record.CreateRecord() {
		utils.FailMessage(c, "create record error")
		return
	}
	utils.SuccessMessage(c, "create record success")
}

type RecordDeleteReq struct {
	RecordId int `json:"record_id" binding:"required"`
}

// RecordDelete 删除账本Ï
func RecordDelete(c *gin.Context) {
	req := RecordDeleteReq{}
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.FailMessage(c, "parse param error")
		return
	}

	record := models.Record{ID: req.RecordId, UserId: c.GetInt("UserId")}
	if !record.DeleteRecord() {
		utils.FailMessage(c, "delete record error")
		return
	}
	utils.SuccessMessage(c, "delete record success")
}

type RecordListReq struct {
	BeginTime string `json:"begin_time" binding:"required"` // 开始时间
	EndTime   string `json:"end_time" binding:"required"`   // 截止时间
}

// RecordList 查询账单
func RecordList(c *gin.Context) {
	req := RecordListReq{}
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.FailMessage(c, "parse param error")
		return
	}
	beginTime, err := time.Parse(utils.TimeLayout, req.BeginTime)
	if err != nil {
		utils.FailMessage(c, "parse begin_time error.")
		return
	}
	endTime, err := time.Parse(utils.TimeLayout, req.EndTime)
	if err != nil {
		utils.FailMessage(c, "parse end_time error.")
		return
	}

	var records []models.Record
	record := models.Record{UserId: c.GetInt("UserId")}
	if !record.ListRecordByTime(beginTime, endTime, &records) {
		utils.FailMessage(c, "list record error")
		return
	}
	res := record.ToRecordView(&records)
	utils.Success(c, gin.H{"records": res}, "list records success")
}

// RecordListAll 查询所有账单
func RecordListAll(c *gin.Context) {
	var records []models.Record
	record := models.Record{UserId: c.GetInt("UserId")}
	if !record.ListRecords(&records) {
		utils.FailMessage(c, "list record error")
		return
	}
	res := record.ToRecordView(&records)
	utils.Success(c, gin.H{"records": res}, "list records success")
}

File: frontend/src/api/record.ts
import axios from './instance/bff_instance'
import { RecordId, RecordItem } from './interface';


const recordApi = {
  listRecord,
  listAllRecords,
  recordCreate,
  deleteRecord,
}

/**
 * 根据时间范围查询账单
 */
export async function listRecord(begin_time: string, end_time: string): Promise<RecordItem[]> {
  try {
    const response = await axios.post<{ records: RecordItem[] }>('/record/list', {
      begin_time: begin_time,
      end_time: end_time,
    });
    return response.data.records;
  } catch (error) {
    console.error('list records error', error)
    return [];
  }
}

/**
 * 查询所有账单
 */
export async function listAllRecords(): Promise<RecordItem[]> {
  try {
    const response = await axios.post<{ records: RecordItem[] }>('/record/list_all');
    return response.data.records;
  } catch (error) {
    console.error('list records error', error)
    return [];
  }
}

export interface RecordCreateParams {
  classification_id?: number, // 分类ID
  sub_classification_id?: number, // 子分类ID
  note?: string, // 备注
  type?: string, // 账单类型
  amount?: string, // 金额
  consumption_time?: string, // 账单时间
}
/**
 * 创建账单
 * @param params 
 * @returns 
 */
function recordCreate(params: RecordCreateParams) {
  const { classification_id, sub_classification_id, note, type, amount, consumption_time } = params || {};

  return axios({
    method: 'POST',
    url: '/record/create',
    data: {
      classification_id: classification_id, sub_classification_id: sub_classification_id,
      note: note, type: type, amount: amount, consumption_time: consumption_time
    },
  });
}

export interface RecordDeleteParams {
  record_id?: number; // 账单ID
}

/**
 * 删除账单
 */
export async function deleteRecord(record_id: RecordId): Promise<boolean> {
  try {
    await axios.post<any>('/record/delete', {
      record_id: record_id
    });
    return true;
  } catch (error) {
    console.error('list records error', error)
    return false;
  }
}

export default recordApi